int32 left
int32 right
int32 seq_num